lib/mapObjects/CRewardableObject.cpp

```diff
@@ -448,7 +450,7 @@ void CRewardInfo::loadComponents(std::vector<Component> & comps,
 		comps.emplace_back(Component::EXPERIENCE, 1, heroLevel, 0);
 
 	if (manaDiff || manaPercentage >= 0)
-		comps.emplace_back(Component::PRIM_SKILL, 5, manaDiff, 0);
+		comps.emplace_back(Component::PRIM_SKILL, 5, manaDiff ? manaDiff : h->manaLimit() * manaPercentage / 100, 0);
```

Comment:
This is fix for mana well UI bug? Where it says it gives 0 mana?
Correct solution would be to extract mana amount calculations from `CRewardableObject::grantRewardAfterLevelup` method into a separate one & use it here.